File: src/loom/deps.py
"""Dependency operations: cycle detection, blockers, ready, add/remove.

The ``dependencies`` table is derived from each item's ``depends_on``
frontmatter list. Mutations here go through the same write-then-sync
pattern as the rest of :mod:`loom.items` — frontmatter is the source of
truth, the index is rebuilt from it.

Rules (see CLAUDE.md "Non-negotiables"):

* Any non-project item may depend on any other non-project item.
* A dependency is satisfied iff the target's status is the literal
  string ``'done'``. Custom statuses do not satisfy.
* Depending on a project is forbidden; projects have no status to
  satisfy.
"""

from __future__ import annotations

import json
from collections import deque
from collections.abc import Sequence
from dataclasses import dataclass
from pathlib import Path

from .errors import LoomError, NotFound
from .ids import ItemType, parse_qid
from .index import Index, IndexRecord
from .scan import build_record
from .storage import dump, load

DONE_STATUS = "done"
MISSING_TYPE = "missing"


@dataclass(frozen=True, slots=True)
class ItemRef:
    """Lightweight summary of an item for list returns.

    Exists so callers can list ``dependencies()`` / ``dependents()``
    without forcing a full file read for each item.
    """

    qualified_id: str
    type: str
    title: str
    status: str | None

    @classmethod
    def from_record(cls, record: IndexRecord) -> ItemRef:
        return cls(
            qualified_id=record.qualified_id,
            type=record.type,
            title=record.title,
            status=record.status,
        )


# ---------------------------------------------------------------------------
# Read-side computations
# ---------------------------------------------------------------------------


def compute_dependencies(idx: Index, qualified_id: str) -> list[ItemRef]:
    """Return refs for every target *qualified_id* depends on."""
    return _refs_for(idx, idx.dependencies_of(qualified_id))


def compute_dependents(idx: Index, qualified_id: str) -> list[ItemRef]:
    """Return refs for every source that depends on *qualified_id*."""
    return _refs_for(idx, idx.dependents_of(qualified_id))


def compute_blockers(idx: Index, qualified_id: str) -> list[ItemRef]:
    """Return refs for the not-yet-satisfied dependencies of *qualified_id*.

    Reads declared deps from the source's stored frontmatter (the source
    of truth), not the dependencies table — so a broken dep (target
    deleted out from under the source) is still surfaced as a blocker.
    Missing targets are returned as a placeholder ItemRef with
    type='missing'.
    """
    record = idx.get(qualified_id)
    if record is None:
        raise NotFound(qualified_id)
    out: list[ItemRef] = []
    for dep_qid in _declared_deps(record):
        target = idx.get(dep_qid)
        if target is None:
            out.append(
                ItemRef(
                    qualified_id=dep_qid,
                    type=MISSING_TYPE,
                    title="(missing target)",
                    status=None,
                )
            )
        elif target.status != DONE_STATUS:
            out.append(ItemRef.from_record(target))
    return out


def is_pickable(idx: Index, qualified_id: str) -> bool:
    """True iff status=='ready', not archived, and every declared dep is done.

    A missing dep target counts as a blocker — a broken dep can never
    be satisfied, so the source isn't safe to pick up.
    """
    record = idx.get(qualified_id)
    if record is None:
        raise NotFound(qualified_id)
    if record.archived:
        return False
    if record.status != "ready":
        return False
    return not compute_blockers(idx, qualified_id)


def _declared_deps(record: IndexRecord) -> list[str]:
    """Read this item's declared deps from its stored frontmatter snapshot."""
    fm = json.loads(record.frontmatter_json)
    return [str(x) for x in (fm.get("depends_on") or ())]


def _refs_for(idx: Index, qids: tuple[str, ...]) -> list[ItemRef]:
    """Materialize index records for *qids*, dropping any that are missing.

    A missing target shouldn't normally happen (ON DELETE CASCADE would
    have removed the edge), but we tolerate it rather than crash on a
    half-rebuilt index.
    """
    out: list[ItemRef] = []
    for qid in qids:
        rec = idx.get(qid)
        if rec is not None:
            out.append(ItemRef.from_record(rec))
    return out


# ---------------------------------------------------------------------------
# Cycle detection
# ---------------------------------------------------------------------------


def would_create_cycle(
    idx: Index,
    source_qid: str,
    target_qid: str,
) -> list[str] | None:
    """If adding ``source -> target`` would close a cycle, return the cycle path.

    The path lists nodes in traversal order, starting and ending at
    ``source_qid``. Returns ``None`` if no cycle would form.

    Self-loops are *not* handled here — callers should reject
    ``source == target`` with a dedicated error.
    """
    if source_qid == target_qid:
        # Defensive: callers should reject earlier with a clearer error.
        return [source_qid, source_qid]

    # BFS forward from target through existing dep edges. If we reach
    # source, the new edge source->target would close source -> target ->
    # ... -> source.
    visited: set[str] = set()
    queue: deque[tuple[str, list[str]]] = deque([(target_qid, [target_qid])])
    while queue:
        node, path = queue.popleft()
        if node in visited:
            continue
        visited.add(node)
        for nxt in idx.dependencies_of(node):
            if nxt == source_qid:
                # Cycle found: source -> target -> ... -> nxt(=source)
                return [source_qid, *path, nxt]
            queue.append((nxt, [*path, nxt]))
    return None


# ---------------------------------------------------------------------------
# Mutations
# ---------------------------------------------------------------------------


def add_dependency(root: Path, source_qid: str, target_qid: str) -> None:
    """Edit the source's ``depends_on`` list to include *target_qid*.

    Validates: source exists; target exists; neither is a project;
    they are distinct; no cycle would form. Existing duplicate edges are
    a no-op.
    """
    parse_qid(source_qid)
    parse_qid(target_qid)
    idx = Index(root)

    source = idx.get(source_qid)
    if source is None:
        raise NotFound(source_qid)
    target = idx.get(target_qid)
    if target is None:
        raise NotFound(target_qid)

    if source.type == ItemType.PROJECT.value:
        raise LoomError(f"projects cannot have dependencies (source: {source_qid})")
    if target.type == ItemType.PROJECT.value:
        raise LoomError(f"depending on a project is forbidden (target: {target_qid})")
    if source_qid == target_qid:
        raise LoomError(f"an item cannot depend on itself: {source_qid}")

    if target_qid in source.depends_on:
        return  # idempotent — already present

    cycle = would_create_cycle(idx, source_qid, target_qid)
    if cycle is not None:
        from .errors import CycleError

        raise CycleError(source_qid, target_qid, cycle)

    _rewrite_depends_on(root, source, lambda existing: [*existing, target_qid])


def remove_dependency(root: Path, source_qid: str, target_qid: str) -> None:
    """Strip *target_qid* from the source's ``depends_on`` list.

    Raises :class:`NotFound` if the edge is not present — silent no-op
    would mask typos.
    """
    parse_qid(source_qid)
    parse_qid(target_qid)
    idx = Index(root)

    source = idx.get(source_qid)
    if source is None:
        raise NotFound(source_qid)
    if target_qid not in source.depends_on:
        raise NotFound(f"{source_qid} -> {target_qid}")

    _rewrite_depends_on(root, source, lambda existing: [t for t in existing if t != target_qid])


def _rewrite_depends_on(
    root: Path,
    source: IndexRecord,
    transform,
) -> None:
    """Apply *transform* to the source's ``depends_on`` list and resync."""
    from datetime import UTC, datetime

    path = root / source.file_path
    fm, body = load(path)
    existing = [str(x) for x in (fm.get("depends_on") or ())]
    new_list = list(transform(existing))
    if new_list:
        fm["depends_on"] = new_list
    else:
        fm.pop("depends_on", None)
    fm["updated_at"] = datetime.now(UTC).isoformat(timespec="seconds")
    dump(path, fm, body)
    record = build_record(path, root)
    Index(root).apply_record(record)


# ---------------------------------------------------------------------------
# Subtree close
# ---------------------------------------------------------------------------


def topological_sort(
    records: Sequence[IndexRecord],
    idx: Index,
) -> list[IndexRecord]:
    """Return *records* in topological order.

    Two records X and Y where X depends on Y: Y precedes X. Records at the
    same dep-rank are sorted by qualified_id (stable, deterministic).

    Records not present in the input set are ignored as deps (a record may
    have dependencies on items outside the input slice).
    """
    qid_set = {r.qualified_id for r in records}
    by_qid = {r.qualified_id: r for r in records}

    # Build in-degree (incoming-edge count for each record), restricted
    # to deps within the input slice.
    in_degree: dict[str, int] = {qid: 0 for qid in qid_set}
    deps_map: dict[str, list[str]] = {qid: [] for qid in qid_set}
    for qid in qid_set:
        for ref in compute_dependencies(idx, qid):
            if ref.qualified_id in qid_set:
                in_degree[qid] += 1
                deps_map[ref.qualified_id].append(qid)

    # Kahn's algorithm with deterministic ordering at each rank.
    ready = sorted(qid for qid, deg in in_degree.items() if deg == 0)
    out: list[IndexRecord] = []
    while ready:
        next_ready: list[str] = []
        for qid in ready:
            out.append(by_qid[qid])
            for dependent in deps_map[qid]:
                in_degree[dependent] -= 1
                if in_degree[dependent] == 0:
                    next_ready.append(dependent)
        ready = sorted(next_ready)
    # Any remaining items had a cycle; deterministic fallback by qid.
    if len(out) != len(records):
        emitted = {o.qualified_id for o in out}
        out.extend(
            sorted(
                (r for r in records if r.qualified_id not in emitted),
                key=lambda r: r.qualified_id,
            )
        )
    return out


def descendants(idx: Index, parent_qid: str) -> list[IndexRecord]:
    """Return every item below *parent_qid* in the hierarchy.

    Uses Python-side prefix matching to avoid SQL ``LIKE`` wildcard
    issues with project names that contain ``_``.
    """
    project = parent_qid.split(":", 1)[0]
    prefix = parent_qid + ":"
    return [r for r in idx.find(project=project) if r.qualified_id.startswith(prefix)]

File: web/lib/deps.ts
/**
 * Dependency operations — read-side and write-side.
 *
 * Mirrors src/loom/deps.py.
 *
 * Rules:
 * - Only the literal string 'done' satisfies a dependency.
 * - computeBlockers reads deps from stored frontmatter (source of truth),
 *   not the dependencies table, so broken deps are still surfaced.
 * - A missing dep target counts as a blocker (a broken dep can never
 *   be satisfied).
 * - Depending on a project is forbidden.
 * - Cycles are rejected at add time.
 */

import * as path from "path";
import { Index, IndexRecord } from "./index";
import { NotFound, LoomError, CycleError } from "./errors";
import { load, dump } from "./storage";
import { buildRecord } from "./scan";
import { syncOne } from "./rebuild";
import { parse_qid } from "./ids";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

export const DONE_STATUS = "done";
export const MISSING_TYPE = "missing";

// ---------------------------------------------------------------------------
// ItemRef
// ---------------------------------------------------------------------------

/**
 * Lightweight summary of an item for list returns.
 * Mirrors deps.py ItemRef dataclass.
 */
export class ItemRef {
  readonly qualifiedId: string;
  readonly type: string;
  readonly title: string;
  readonly status: string | null;

  constructor(qualifiedId: string, type: string, title: string, status: string | null) {
    this.qualifiedId = qualifiedId;
    this.type = type;
    this.title = title;
    this.status = status;
  }

  static fromRecord(record: IndexRecord): ItemRef {
    return new ItemRef(record.qualified_id, record.type, record.title, record.status);
  }
}

// ---------------------------------------------------------------------------
// Read-side helpers
// ---------------------------------------------------------------------------

/** Read the declared depends_on list from a record's stored frontmatter. */
function declaredDeps(record: IndexRecord): string[] {
  try {
    const fm = JSON.parse(record.frontmatter_json) as Record<string, unknown>;
    const deps = fm["depends_on"];
    if (!Array.isArray(deps)) return [];
    return deps.map(String);
  } catch {
    return [];
  }
}

/** Materialize index records for qids, dropping any that are missing. */
function refsFor(idx: Index, qids: readonly string[]): ItemRef[] {
  const out: ItemRef[] = [];
  for (const qid of qids) {
    const rec = idx.get(qid);
    if (rec !== null) {
      out.push(ItemRef.fromRecord(rec));
    }
  }
  return out;
}

// ---------------------------------------------------------------------------
// Public read-side API
// ---------------------------------------------------------------------------

/**
 * Return refs for every target that qualifiedId depends on.
 * Uses the DB dependencies table (index-derived; may be empty for broken deps).
 */
export function computeDependencies(idx: Index, qualifiedId: string): ItemRef[] {
  return refsFor(idx, idx.dependenciesOf(qualifiedId));
}

/**
 * Return refs for every source that depends on qualifiedId.
 */
export function computeDependents(idx: Index, qualifiedId: string): ItemRef[] {
  return refsFor(idx, idx.dependentsOf(qualifiedId));
}

/**
 * Return refs for the not-yet-satisfied dependencies of qualifiedId.
 *
 * Reads declared deps from frontmatter (the source of truth), so a broken
 * dep (target deleted) is still surfaced as a blocker with type='missing'.
 * Missing targets are returned as placeholder ItemRefs.
 */
export function computeBlockers(idx: Index, qualifiedId: string): ItemRef[] {
  const record = idx.get(qualifiedId);
  if (record === null) {
    throw new NotFound(qualifiedId);
  }
  const out: ItemRef[] = [];
  for (const depQid of declaredDeps(record)) {
    const target = idx.get(depQid);
    if (target === null) {
      out.push(new ItemRef(depQid, MISSING_TYPE, "(missing target)", null));
    } else if (target.status !== DONE_STATUS) {
      out.push(ItemRef.fromRecord(target));
    }
  }
  return out;
}

/**
 * True iff status=='ready', not archived, and every declared dep is done.
 *
 * A missing dep target counts as a blocker — a broken dep can never
 * be satisfied, so the source isn't safe to pick up.
 */
export function isPickable(idx: Index, qualifiedId: string): boolean {
  const record = idx.get(qualifiedId);
  if (record === null) {
    throw new NotFound(qualifiedId);
  }
  if (record.archived) return false;
  if (record.status !== "ready") return false;
  return computeBlockers(idx, qualifiedId).length === 0;
}

// ---------------------------------------------------------------------------
// Cycle detection
// ---------------------------------------------------------------------------

/**
 * If adding source -> target would close a cycle, return the cycle path.
 *
 * The path lists nodes in traversal order, starting and ending at source.
 * Returns null if no cycle would form.
 *
 * Self-loops are handled defensively: source == target returns a 2-element path.
 */
export function wouldCreateCycle(
  idx: Index,
  sourceQid: string,
  targetQid: string
): string[] | null {
  if (sourceQid === targetQid) {
    return [sourceQid, sourceQid];
  }

  // BFS forward from target through existing dep edges.
  // If we reach source, the new edge source->target would close a cycle.
  const visited = new Set<string>();
  const queue: Array<{ node: string; path: string[] }> = [
    { node: targetQid, path: [targetQid] },
  ];

  while (queue.length > 0) {
    const entry = queue.shift()!;
    const { node, path } = entry;
    if (visited.has(node)) continue;
    visited.add(node);

    for (const nxt of idx.dependenciesOf(node)) {
      if (nxt === sourceQid) {
        // Cycle found: source -> target -> ... -> nxt(=source)
        return [sourceQid, ...path, nxt];
      }
      queue.push({ node: nxt, path: [...path, nxt] });
    }
  }

  return null;
}

// ---------------------------------------------------------------------------
// Write-side helpers
// ---------------------------------------------------------------------------

/** Now-ISO string for updated_at bumps. */
function nowIso(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "+00:00");
}

/**
 * Apply transform to the source's depends_on list, rewrite the file,
 * and re-sync the index.
 */
async function rewriteDependsOn(
  root: string,
  source: IndexRecord,
  transform: (existing: string[]) => string[]
): Promise<void> {
  const filePath = path.join(root, source.file_path);
  const [fm, body] = load(filePath);
  const existing = Array.isArray(fm["depends_on"])
    ? (fm["depends_on"] as unknown[]).map(String)
    : [];
  const newList = transform(existing);
  if (newList.length > 0) {
    fm["depends_on"] = newList;
  } else {
    delete fm["depends_on"];
  }
  fm["updated_at"] = nowIso();
  dump(filePath, fm, body);
  const record = await buildRecord(filePath, root);
  new Index(root).applyRecord(record);
}

// ---------------------------------------------------------------------------
// Write-side API
// ---------------------------------------------------------------------------

/**
 * Add a dependency edge: source depends on target.
 *
 * Validates: both exist; neither is a project; they are distinct;
 * no cycle would form. Idempotent if edge already exists.
 */
export async function addDependency(
  root: string,
  sourceQid: string,
  targetQid: string
): Promise<void> {
  parse_qid(sourceQid);
  parse_qid(targetQid);
  const idx = new Index(root);

  const source = idx.get(sourceQid);
  if (source === null) throw new NotFound(sourceQid);
  const target = idx.get(targetQid);
  if (target === null) throw new NotFound(targetQid);

  if (source.type === "project") {
    throw new LoomError(`projects cannot have dependencies (source: ${sourceQid})`);
  }
  if (target.type === "project") {
    throw new LoomError(`depending on a project is forbidden (target: ${targetQid})`);
  }
  if (sourceQid === targetQid) {
    throw new LoomError(`an item cannot depend on itself: ${sourceQid}`);
  }

  // Idempotent: already present
  if (source.depends_on.includes(targetQid)) return;

  const cycle = wouldCreateCycle(idx, sourceQid, targetQid);
  if (cycle !== null) {
    throw new CycleError(sourceQid, targetQid, cycle);
  }

  await rewriteDependsOn(root, source, (existing) => [...existing, targetQid]);
}

/**
 * Remove a dependency edge from source's depends_on list.
 *
 * Raises NotFound if the source doesn't exist or the edge is absent.
 */
export async function removeDependency(
  root: string,
  sourceQid: string,
  targetQid: string
): Promise<void> {
  parse_qid(sourceQid);
  parse_qid(targetQid);
  const idx = new Index(root);

  const source = idx.get(sourceQid);
  if (source === null) throw new NotFound(sourceQid);
  if (!source.depends_on.includes(targetQid)) {
    throw new NotFound(`${sourceQid} -> ${targetQid}`);
  }

  await rewriteDependsOn(root, source, (existing) =>
    existing.filter((t) => t !== targetQid)
  );
}

// ---------------------------------------------------------------------------
// Topological sort + descendants
// ---------------------------------------------------------------------------

/**
 * Return records in topological dep-order (dependencies first).
 *
 * Records at the same dep-rank are sorted by qualified_id for determinism.
 * Deps on items outside the input set are ignored.
 * If a cycle exists among the input set, remaining items are appended sorted.
 *
 * Mirrors deps.py topological_sort.
 */
export function topologicalSort(
  records: IndexRecord[],
  idx: Index
): IndexRecord[] {
  if (records.length === 0) return [];

  const qidSet = new Set(records.map((r) => r.qualified_id));
  const byQid = new Map(records.map((r) => [r.qualified_id, r]));

  // Build in-degree and adjacency restricted to input set.
  const inDegree = new Map<string, number>();
  const depsMap = new Map<string, string[]>(); // qid -> list of dependents

  for (const qid of qidSet) {
    inDegree.set(qid, 0);
    depsMap.set(qid, []);
  }
  for (const qid of qidSet) {
    for (const dep of idx.dependenciesOf(qid)) {
      if (qidSet.has(dep)) {
        inDegree.set(qid, (inDegree.get(qid) ?? 0) + 1);
        depsMap.get(dep)!.push(qid);
      }
    }
  }

  // Kahn's algorithm with deterministic ordering at each rank.
  let ready = [...qidSet]
    .filter((qid) => (inDegree.get(qid) ?? 0) === 0)
    .sort();
  const out: IndexRecord[] = [];

  while (ready.length > 0) {
    const nextReady: string[] = [];
    for (const qid of ready) {
      out.push(byQid.get(qid)!);
      for (const dependent of depsMap.get(qid) ?? []) {
        const newDeg = (inDegree.get(dependent) ?? 0) - 1;
        inDegree.set(dependent, newDeg);
        if (newDeg === 0) nextReady.push(dependent);
      }
    }
    ready = nextReady.sort();
  }

  // Fallback for cyclic remainder (sorted by qid for determinism).
  if (out.length !== records.length) {
    const emitted = new Set(out.map((r) => r.qualified_id));
    const remaining = records
      .filter((r) => !emitted.has(r.qualified_id))
      .sort((a, b) => a.qualified_id.localeCompare(b.qualified_id));
    out.push(...remaining);
  }

  return out;
}

/**
 * Return every item below parentQid in the hierarchy.
 *
 * Uses prefix matching: any item whose qualified_id starts with
 * parentQid + ":" is a descendant.
 *
 * Mirrors deps.py descendants.
 */
export function descendants(idx: Index, parentQid: string): IndexRecord[] {
  const project = parentQid.split(":")[0]!;
  const prefix = parentQid + ":";
  const all = idx.find({ project });
  return all.filter((r) => r.qualified_id.startsWith(prefix));
}
